lib/chef/provider.rb

```diff
@@ -376,7 +376,18 @@ module ClassMethods
         # compile the resources, converging them, and then checking if any
         # were updated (and updating new-resource if so)
         def action(name, &block)
-          define_method("action_#{name}") { compile_and_converge_action(&block) }
```

Comment:
what if you pull the implementation of compile_and_converge_action into the defined method?  also does wrapping the block in a Proc or a lambda help at all?